File: frontend/src/types/payment.ts
/**
 * Payment types for ExtraShifty frontend
 * These types match the backend schemas in /backend/app/schemas/payment.py and /backend/app/schemas/wallet.py
 */

// =============================================================================
// Transaction Types (matches backend TransactionType enum)
// =============================================================================

/**
 * Backend transaction types (from backend/app/schemas/payment.py)
 */
export type BackendTransactionType =
  | 'topup'           // Adding funds to wallet
  | 'reserve'         // Reserving funds for accepted shift
  | 'release'         // Releasing reserved funds back to available
  | 'settlement'      // Paying staff after shift completion
  | 'commission'      // Platform commission fee
  | 'payout'          // Withdrawal to bank account
  | 'refund'          // Refund to company wallet
  | 'cancellation_fee' // Fee for late cancellation
  | 'penalty'         // Penalty deduction

/**
 * UI-friendly transaction types for frontend display
 * These map to backend types but use more user-friendly names
 */
export type UITransactionType =
  | 'earning'         // Staff earnings from shifts
  | 'withdrawal'      // Withdrawal to bank account (maps to 'payout')
  | 'top_up'          // Adding funds (maps to 'topup')
  | 'payment'         // Payment for services

/**
 * Combined transaction type that supports both backend and UI types
 * Use this for type-safe handling across the application
 */
export type TransactionType = BackendTransactionType | UITransactionType

export type TransactionStatus = 'pending' | 'completed' | 'failed' | 'cancelled'

// =============================================================================
// Payout Types (matches backend PayoutStatus and PayoutType enums)
// =============================================================================

export type PayoutStatus = 'pending' | 'in_transit' | 'paid' | 'failed' | 'cancelled'
export type PayoutType = 'weekly' | 'instant'

// =============================================================================
// Payment Method Types (matches backend PaymentMethodType enum)
// =============================================================================

export type PaymentMethodType = 'card' | 'bank_account' | 'ach'

// =============================================================================
// Wallet Types (matches backend WalletType and WalletStatus enums)
// =============================================================================

export type WalletType = 'company' | 'staff' | 'agency' | 'platform'
export type WalletStatus = 'active' | 'grace_period' | 'suspended'

// =============================================================================
// Cancellation Types (matches backend CancellationPolicy and CancelledBy enums)
// =============================================================================

export type CancellationPolicy = 'full_refund' | 'partial_refund' | 'no_refund' | 'worker_compensation'
export type CancelledBy = 'company' | 'worker' | 'platform'

// =============================================================================
// Wallet Balance Response (matches backend BalanceResponse)
// =============================================================================

export interface WalletBalance {
  wallet_id?: number
  available: number
  reserved: number
  pending_payout?: number
  total: number
  currency: string
  low_balance_threshold?: number
  is_low_balance?: boolean
}

// =============================================================================
// Top-up Request/Response (matches backend TopupRequest/TopupResponse)
// =============================================================================

export interface TopupRequest {
  amount: number
  payment_method_id: number
  idempotency_key?: string
}

export interface TopupResponse {
  transaction_id: number
  amount: number
  new_balance: number
  status: string
  message: string
}

// =============================================================================
// Auto Top-up Configuration (matches backend AutoTopupConfigRequest/Response)
// =============================================================================

export interface AutoTopupConfig {
  enabled: boolean
  threshold?: number
  topup_amount?: number
  payment_method_id?: number
}

export interface AutoTopupConfigResponse extends AutoTopupConfig {
  message: string
}

// =============================================================================
// Reserve Request/Response (matches backend ReserveRequest/ReserveResponse)
// =============================================================================

export interface ReserveRequest {
  idempotency_key?: string
}

export interface ReserveResponse {
  hold_id: number
  shift_id: number
  amount_reserved: number
  remaining_balance: number
  expires_at: string
  message: string
}

// =============================================================================
// Insufficient Funds Response (matches backend InsufficientFundsResponse)
// =============================================================================

export interface InsufficientFundsError {
  error: 'insufficient_funds'
  required_amount: number
  available_amount: number
  shortfall: number
  minimum_balance?: number
  message: string
}

// =============================================================================
// Settlement Types (matches backend SettlementSplit/SettlementResponse)
// =============================================================================

export interface SettlementSplit {
  gross_amount: number
  platform_fee: number
  platform_fee_rate: number
  worker_amount: number
  agency_fee?: number
}

export interface SettlementResponse {
  shift_id: number
  settlement_id: number
  actual_hours: number
  gross_amount: number
  split: SettlementSplit
  transactions: Array<{
    transaction_id: number
    type: string
    amount: string
    fee?: string
    net_amount?: string
  }>
  message: string
}

// =============================================================================
// Cancellation Types (matches backend CancellationRequest/CancellationResponse)
// =============================================================================

export interface CancellationRequest {
  cancelled_by: CancelledBy
  reason?: string
  idempotency_key?: string
}

export interface CancellationResponse {
  shift_id: number
  cancelled_by: CancelledBy
  policy_applied: CancellationPolicy
  refund_amount: number
  worker_compensation: number
  transactions: Array<{
    transaction_id: number
    type: string
    amount: string
  }>
  message: string
}

// =============================================================================
// Payout Types (matches backend PayoutRequest/PayoutResponse)
// =============================================================================

export interface PayoutRequest {
  amount?: number
  idempotency_key?: string
}

export interface PayoutResponse {
  payout_id: number
  amount: number
  fee: number
  net_amount: number
  status: PayoutStatus
  estimated_arrival?: string
  message: string
}

// =============================================================================
// Payout Schedule (matches backend PayoutScheduleResponse)
// =============================================================================

export interface PayoutScheduleItem {
  scheduled_date: string
  estimated_amount: number
  status: string
}

export interface PayoutScheduleResponse {
  next_payout_date?: string
  minimum_threshold: number
  current_balance: number
  scheduled_payouts: PayoutScheduleItem[]
}

// =============================================================================
// Payout History (matches backend PayoutHistoryItem/PayoutHistoryResponse)
// =============================================================================

export interface PayoutHistoryItem {
  payout_id: number
  amount: number
  fee: number
  net_amount: number
  status: PayoutStatus
  payout_type: PayoutType
  created_at: string
  completed_at?: string
}

export interface PayoutHistoryResponse {
  items: PayoutHistoryItem[]
  total: number
}

// =============================================================================
// Minimum Balance (matches backend MinimumBalanceRequest/MinimumBalanceResponse)
// =============================================================================

export interface MinimumBalanceRequest {
  minimum_balance: number
}

export interface MinimumBalanceResponse {
  wallet_id: number
  minimum_balance: number
  available_balance: number
  message: string
}

// =============================================================================
// Wallet Status Response
// =============================================================================

export interface WalletStatusResponse {
  wallet_id: number
  status: WalletStatus
  is_active: boolean
  can_accept_shifts: boolean
  grace_period_ends_at?: string
  last_failed_topup_at?: string
  suspension_reason?: string
}

// =============================================================================
// Payment Method (matches backend PaymentMethodRead)
// =============================================================================

export interface PaymentMethod {
  id: number
  user_id: number
  type: PaymentMethodType
  last_four: string
  brand?: string
  is_default: boolean
  created_at: string
}

// =============================================================================
// Transaction (matches backend TransactionRead)
// =============================================================================

export interface Transaction {
  id: number
  wallet_id: number
  type: TransactionType
  amount: number
  description: string
  status: TransactionStatus
  reference_id?: string
  created_at: string
}

// =============================================================================
// Wallet (matches backend WalletRead)
// =============================================================================

export interface Wallet {
  id: number
  user_id: number
  balance: number
  currency: string
  created_at: string
  updated_at: string
}

File: backend/app/schemas/payment.py
"""Payment schemas for ExtraShifty payment flow."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from ..models.payment import PayoutStatus
from ..models.wallet import PaymentMethodType


class TopupRequest(BaseModel):
    """Request schema for wallet topup."""

    amount: Decimal = Field(gt=0, description="Amount to add to wallet")
    payment_method_id: int = Field(description="Payment method to charge")
    idempotency_key: str | None = Field(
        default=None,
        max_length=255,
        description="Unique key to prevent duplicate charges",
    )


class TopupResponse(BaseModel):
    """Response schema for wallet topup."""

    transaction_id: int
    amount: Decimal
    new_balance: Decimal
    status: str
    message: str


class AutoTopupConfigRequest(BaseModel):
    """Request schema for auto-topup configuration."""

    enabled: bool = Field(description="Enable or disable auto-topup")
    threshold: Decimal | None = Field(
        default=None,
        ge=0,
        description="Balance threshold that triggers auto-topup",
    )
    topup_amount: Decimal | None = Field(
        default=None,
        gt=0,
        description="Amount to add when auto-topup triggers",
    )
    payment_method_id: int | None = Field(
        default=None,
        description="Payment method to use for auto-topup",
    )


class AutoTopupConfigResponse(BaseModel):
    """Response schema for auto-topup configuration."""

    enabled: bool
    threshold: Decimal | None
    topup_amount: Decimal | None
    payment_method_id: int | None
    message: str


class BalanceResponse(BaseModel):
    """Response schema for wallet balance."""

    model_config = ConfigDict(from_attributes=True)

    wallet_id: int
    available: Decimal = Field(description="Available balance for new reservations")
    reserved: Decimal = Field(description="Funds held for active shifts")
    pending_payout: Decimal = Field(description="Earnings pending payout")
    total: Decimal = Field(description="Total balance (available + reserved)")
    currency: str = Field(default="EUR")


class ReserveRequest(BaseModel):
    """Request schema for reserving shift funds."""

    idempotency_key: str | None = Field(
        default=None,
        max_length=255,
        description="Unique key to prevent duplicate reservations",
    )


class ReserveResponse(BaseModel):
    """Response schema for fund reservation."""

    hold_id: int
    shift_id: int
    amount_reserved: Decimal
    remaining_balance: Decimal
    expires_at: datetime
    message: str


class InsufficientFundsResponse(BaseModel):
    """Response for insufficient funds (HTTP 402)."""

    error: str = "insufficient_funds"
    required_amount: Decimal
    available_amount: Decimal
    shortfall: Decimal
    minimum_balance: Decimal | None = None
    message: str


class MinimumBalanceRequest(BaseModel):
    """Request schema for setting minimum balance."""

    minimum_balance: Decimal = Field(
        ge=0,
        description="Minimum balance that must remain after accepting a shift",
    )


class MinimumBalanceResponse(BaseModel):
    """Response schema for minimum balance update."""

    wallet_id: int
    minimum_balance: Decimal
    available_balance: Decimal
    message: str


class SettlementSplit(BaseModel):
    """Details of settlement split."""

    gross_amount: Decimal
    platform_fee: Decimal
    platform_fee_rate: Decimal
    worker_amount: Decimal
    agency_fee: Decimal | None = None


class SettlementResponse(BaseModel):
    """Response schema for shift settlement."""

    shift_id: int
    settlement_id: int
    actual_hours: Decimal
    gross_amount: Decimal
    split: SettlementSplit
    transactions: list[dict]
    message: str


class CancellationPolicy(str, Enum):
    """Cancellation policy types."""

    FULL_REFUND = "full_refund"
    PARTIAL_REFUND = "partial_refund"
    NO_REFUND = "no_refund"
    WORKER_COMPENSATION = "worker_compensation"


class CancelledBy(str, Enum):
    """Who cancelled the shift."""

    COMPANY = "company"
    WORKER = "worker"
    PLATFORM = "platform"


class CancellationRequest(BaseModel):
    """Request schema for shift cancellation."""

    cancelled_by: CancelledBy
    reason: str | None = Field(default=None, max_length=500)
    idempotency_key: str | None = Field(
        default=None,
        max_length=255,
        description="Unique key to prevent duplicate cancellations",
    )


class CancellationResponse(BaseModel):
    """Response schema for shift cancellation."""

    shift_id: int
    cancelled_by: CancelledBy
    policy_applied: CancellationPolicy
    refund_amount: Decimal
    worker_compensation: Decimal
    transactions: list[dict]
    message: str


class PayoutRequest(BaseModel):
    """Request schema for instant payout."""

    amount: Decimal | None = Field(
        default=None,
        gt=0,
        description="Amount to withdraw (None = full available balance)",
    )
    idempotency_key: str | None = Field(
        default=None,
        max_length=255,
        description="Unique key to prevent duplicate payouts",
    )


class PayoutResponse(BaseModel):
    """Response schema for payout."""

    model_config = ConfigDict(from_attributes=True)

    payout_id: int
    amount: Decimal
    fee: Decimal
    net_amount: Decimal
    status: PayoutStatus
    estimated_arrival: datetime | None = None
    message: str


class PayoutScheduleItem(BaseModel):
    """Single scheduled payout item."""

    scheduled_date: datetime
    estimated_amount: Decimal
    status: str


class PayoutScheduleResponse(BaseModel):
    """Response schema for payout schedule."""

    next_payout_date: datetime | None
    minimum_threshold: Decimal
    current_balance: Decimal
    scheduled_payouts: list[PayoutScheduleItem]


class PayoutHistoryItem(BaseModel):
    """Single payout history item."""

    model_config = ConfigDict(from_attributes=True)

    payout_id: int
    amount: Decimal
    fee: Decimal
    net_amount: Decimal
    status: PayoutStatus
    payout_type: str  # "instant" or "weekly"
    created_at: datetime
    completed_at: datetime | None = None


class PayoutHistoryResponse(BaseModel):
    """Response schema for payout history."""

    items: list[PayoutHistoryItem]
    total: int
